### ISSUE TITLE
Implementation question regarding PSA and monte carlo samples

### ISSUE THREAD

**Issue by selflein** [hehaodele/ProbGAN#3]
Hi, I am really liking the paper and implementation!
I have few questions regarding implementation details:
#### Partial Sum Approximation (PSA) implementation 
Do I understand that right that this code
```
with torch.no_grad():
       self.ds_hist_avg = nn.Conv2d(self.ndf * 8, num_discs, 3, 1, 0, bias=False)
       self.len_hist = 1.0
``` 
in `multi_discriminator_48_with_history` stores the weights for the patial sum approximation?
Expanding on that it seems that one does not want update these weights through gradient descent but based on the weights of the current iteration. Is this correct?
When experimenting with this code I found that the implementation above does not actually prevent PyTorch from creating gradients:
```
In [11]: with torch.no_grad(): 
    ...:     a = torch.nn.Linear(32, 32) 
    ...:                                                                                                                                                                                                                                                                       
In [12]: a.weight.requires_grad                                                                                                                                                                                                                                                
Out[12]: True
```

#### Splitting batch 
I noticed that every batch is split into `n` parts (where `n` is the number of generator samples) which are then feed throught the one generator sample and then through one discriminator. So every discriminator will always get samples from the same subset of generators. 

Is this crucial from a theoretical perspective or would feeding the entire batch through every generator and discriminator provide the same results? (In my experiments it does not but I would appreciate your opinion on this.)

**Comment by hehaodele** [comment]
Hi Sven,

Thanks for your comments!

#### Weights for past discriminators
You are correct that the past discriminators' weights are supposed to be untrainable. It seems that it was my fault that the current implementation did not prevent from creating the gradient. Thanks for finding the bug! I will fix it as soon as possible.

#### Splitting batch
It is correct in a batch of generated data, every generator will generate 1/n of them.
Then for the discriminators, they actually are feed the whole batch data which means every discriminator sees the fake data from all generators. So it should not be true that ''every discriminator will always get samples from the same subset of generators.''

**Comment by selflein** [comment]
Thank you for your fast response! I feel like PyTorch is not really intuitive in using the `torch.no_grad()` context manager so this is an easy mistake to make. 
Anyway, your answer clears up my remaining confusions. Thank you!
